src/engine/ExportQueryExecutionTrees.cpp

```diff
@@ -83,31 +85,82 @@ ExportQueryExecutionTrees::getIdTables(const Result& result) {
+  uint64_t effectiveLimit = limitOffset.limitOrDefault();
+  uint64_t effectiveExportLimit = limitOffset.exportLimitOrDefault();
+
+  // Make sure that the export limit is at most the limit (increasing the
+  // export limit beyond the limit has no effect).
+  effectiveExportLimit = std::min(effectiveExportLimit, effectiveLimit);
+
+  // Iterate over the result in blocks.
   for (TableConstRefWithVocab& tableWithVocab : getIdTables(result)) {
-    uint64_t currentOffset =
-        limitOffset.actualOffset(tableWithVocab.idTable_.numRows());
-    uint64_t upperBound =
-        limitOffset.upperBound(tableWithVocab.idTable_.numRows());
-    if (currentOffset != upperBound) {
-      co_yield {std::move(tableWithVocab),
-                std::views::iota(currentOffset, upperBound)};
+    // If all rows in the current block are before the effective offset, we can
+    // skip the block entirely. If not, there is at least something to count
+    // and maybe also something to yield.
+    uint64_t currentBlockSize = tableWithVocab.idTable_.numRows();
+    if (effectiveOffset >= currentBlockSize) {
+      effectiveOffset -= currentBlockSize;
+      continue;
     }
-    limitOffset._offset -= currentOffset;
-    if (limitOffset._limit.has_value()) {
-      limitOffset._limit =
-          limitOffset._limit.value() - (upperBound - currentOffset);
+    AD_CORRECTNESS_CHECK(effectiveOffset < currentBlockSize);
+    AD_CORRECTNESS_CHECK(effectiveLimit > 0);
+
+    // Compute the range of rows to be exported (can by zero) and to be counted
+    // (always non-zero at this point).
+    uint64_t rangeBegin = effectiveOffset;
+    uint64_t numRowsToBeExported =
+        std::min(effectiveExportLimit, currentBlockSize - rangeBegin);
+    uint64_t numRowsToBeCounted =
+        std::min(effectiveLimit, currentBlockSize - rangeBegin);
+    AD_CORRECTNESS_CHECK(rangeBegin + numRowsToBeExported <= currentBlockSize);
+    AD_CORRECTNESS_CHECK(rangeBegin + numRowsToBeCounted <= currentBlockSize);
+    AD_CORRECTNESS_CHECK(numRowsToBeCounted > 0);
+
+    // If there is something to be exported, yield it.
+    if (numRowsToBeExported > 0) {
+      co_yield {std::move(tableWithVocab),
+                std::views::iota(rangeBegin, rangeBegin + numRowsToBeExported)};
     }
-    if (limitOffset._limit.value_or(1) == 0) {
-      break;
+
+    // Add to `resultSize` and update the effective offset (which becomes zero
+    // after the first non-skipped block) and limits (make sure to never go
+    // below zero and `std::numeric_limits<uint64_t>::max()` stays there).
+    resultSize += numRowsToBeCounted;
+    effectiveOffset = 0;
+    auto reduceLimit = [&](uint64_t& limit, uint64_t subtrahend) {
+      if (limit != std::numeric_limits<uint64_t>::max()) {
+        limit = limit > subtrahend ? limit - subtrahend : 0;
+      }
+    };
+    reduceLimit(effectiveLimit, numRowsToBeCounted);
+    reduceLimit(effectiveExportLimit, numRowsToBeCounted);
```

Comment:
Haven't checked, but I could imagine there's a shorter way to write this logic if using the code from the `LimitOffsetClause` class

Reply:
I deliberately did not use the methods from `LimitOffsetClause` here because they are opaque and do not apply directly here. What I did is simpler and more transparent.

I addressed your other comments, thanks a lot for the quick review!